tests/ui/deriving/deriving-coerce-pointee-neg.stderr

```diff
@@ -1,124 +1,143 @@
-  --> $DIR/deriving-coerce-pointee-neg.rs:72:33
+  --> $DIR/deriving-coerce-pointee-neg.rs:73:33
    |
 LL |                     struct UhOh<#[pointee] T>(T);
    |                                 ^^^^^^^^^^
 
 error: the `#[pointee]` attribute may only be used on generic parameters
-  --> $DIR/deriving-coerce-pointee-neg.rs:86:21
+  --> $DIR/deriving-coerce-pointee-neg.rs:87:21
    |
 LL |         struct UhOh<#[pointee] T>(T);
    |                     ^^^^^^^^^^
 
 error: the `#[pointee]` attribute may only be used on generic parameters
-  --> $DIR/deriving-coerce-pointee-neg.rs:101:25
+  --> $DIR/deriving-coerce-pointee-neg.rs:102:25
    |
 LL |             struct UhOh<#[pointee] T>(T);
    |                         ^^^^^^^^^^
 
 error[E0392]: lifetime parameter `'a` is never used
-  --> $DIR/deriving-coerce-pointee-neg.rs:18:16
+  --> $DIR/deriving-coerce-pointee-neg.rs:19:16
    |
 LL | struct NoField<'a, #[pointee] T: ?Sized> {}
    |                ^^ unused lifetime parameter
    |
    = help: consider removing `'a`, referring to it in a field, or using a marker such as `PhantomData`
 
 error[E0392]: type parameter `T` is never used
-  --> $DIR/deriving-coerce-pointee-neg.rs:18:31
+  --> $DIR/deriving-coerce-pointee-neg.rs:19:31
    |
 LL | struct NoField<'a, #[pointee] T: ?Sized> {}
    |                               ^ unused type parameter
    |
    = help: consider removing `T`, referring to it in a field, or using a marker such as `PhantomData`
 
 error[E0392]: lifetime parameter `'a` is never used
-  --> $DIR/deriving-coerce-pointee-neg.rs:25:20
+  --> $DIR/deriving-coerce-pointee-neg.rs:26:20
    |
 LL | struct NoFieldUnit<'a, #[pointee] T: ?Sized>();
    |                    ^^ unused lifetime parameter
    |
    = help: consider removing `'a`, referring to it in a field, or using a marker such as `PhantomData`
 
 error[E0392]: type parameter `T` is never used
-  --> $DIR/deriving-coerce-pointee-neg.rs:25:35
+  --> $DIR/deriving-coerce-pointee-neg.rs:26:35
    |
 LL | struct NoFieldUnit<'a, #[pointee] T: ?Sized>();
    |                                   ^ unused type parameter
    |
    = help: consider removing `T`, referring to it in a field, or using a marker such as `PhantomData`
 
 error[E0802]: `derive(CoercePointee)` is only applicable to `struct` with `repr(transparent)` layout
-  --> $DIR/deriving-coerce-pointee-neg.rs:47:1
+  --> $DIR/deriving-coerce-pointee-neg.rs:48:1
    |
 LL | struct NotTransparent<'a, #[pointee] T: ?Sized> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 error[E0802]: `derive(CoercePointee)` is only applicable to `struct` with `repr(transparent)` layout
-  --> $DIR/deriving-coerce-pointee-neg.rs:140:1
+  --> $DIR/deriving-coerce-pointee-neg.rs:141:1
    |
 LL | struct TryToWipeRepr<'a, #[pointee] T: ?Sized> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-error: aborting due to 17 previous errors
+error[E0277]: the trait bound `Box<T>: Unsize<Box<T {unsized}>>` is not satisfied
```

Comment:
This PR changes introduces quite a lot of machinery just to rename `__S` to `T {unsized}`. I don't actually think that's very different, and not worth the changes.

If we're going to do anything here, then it needs to be about changing the *wording* of the error message to explain what this `Unsize` obligation is about.

Reply:
I would like to quote #134846 that this PR attempts to fix:

> Desired output
> Something that does not mention internal things like `Unsize` or `__S`

So I am thinking, should we instead redact all errors from unfulfilled `Unsize` obligation entirely and replace them with messages to say `unsizing <some type> is not permitted`, or something along this line?

Concretely, if we encounter fulfillment error for `Ty1: Unsize<Ty2>`, we write out a message saying `unsizing Ty1 is not permitted` and forget about `Ty2`, since `Ty2` may contain internal stuffs like `__S` in most cases. Is this what we are going for here?

Reply:
Sure. I think all this PR needs to do is intercept the errors in:

https://github.com/rust-lang/rust/blob/06a24e98c612d72add1c3355282510513b153dcc/compiler/rustc_hir_analysis/src/coherence/builtin.rs#L560-L562

And:

https://github.com/rust-lang/rust/blob/06a24e98c612d72add1c3355282510513b153dcc/compiler/rustc_hir_analysis/src/coherence/builtin.rs#L348-L350

And just say something like: "In order for `ADT` to have a valid implementation of `CoerceUnsized`, you need to be able to unsize `&W<T>`."

Most importantly, it needs to explain *why* the unsize isn't valid. I know you're interested in fixing what was requested in #134846, but please think about the bigger picture, and also please think about minimizing the changes to the compiler while trying to fix the issue.